tools/nl/embeddings/utils.py

```diff
@@ -124,84 +75,123 @@ def get_model(catalog: Catalog, env: Env, model_name: str) -> EmbeddingsModel:
   return model
 
 
-def build_and_save_preindex(fm: FileManager) -> Tuple[List[str], List[str]]:
+def get_saved_embeddings(embeddings_path: str) -> List[SentenceObject]:
+  """Load saved embeddings from a CSV file."""
+  try:
+    if gcs.is_gcs_path(embeddings_path):
+      embeddings_path = gcs.maybe_download(embeddings_path)
+    df = pd.read_csv(embeddings_path)
+    sentences = []
+    for _, row in df.iterrows():
+      dcid = row['dcid']
+      sentence = row['sentence']
+      vector = row.drop(labels=['dcid', 'sentence']).astype(float).tolist()
+      sentences.append(SentenceObject(text=sentence, dcid=dcid, vector=vector))
+    return sentences
+  except Exception as e:
+    logging.error(e)
+    return []
+
+
+def build_and_save_preindex(fm: FileManager) -> List[SentenceObject]:
   """
-  Build preindex records (sentence -> dcid) from a directory of CSV files.
+  Build preindex records (text -> dcid) from a directory of CSV files.
   """
   text2sv: Dict[str, set[str]] = {}
   for file_name in glob.glob(fm.local_input_dir() + "/[!_]*.csv"):
     with open(file_name) as f:
       reader = csv.DictReader(f)
       for row in reader:
-        sentences = row[_COL_SENTENCE].split(';')
-        for sentence in sentences:
-          if sentence not in text2sv:
-            text2sv[sentence] = set()
-          text2sv[sentence].add(row[_COL_DCID])
-  texts = []
-  dcids = []
+        texts = row[_COL_SENTENCE].split(';')
+        for text in texts:
+          text = text.strip()
+          if text == '':
+            continue
+          if text not in text2sv:
+            text2sv[text] = set()
+          text2sv[text].add(row[_COL_DCID])
+
+  sentences = [
+      SentenceObject(text, ';'.join(sorted(dcids)), [])
+      for text, dcids in text2sv.items()
+  ]
+  sentences.sort(key=lambda x: x.text)
+
   # Write preindex as CSV
   with open(fm.preindex_csv_path(), 'w') as csvfile:
     csv_writer = csv.writer(csvfile, delimiter=',')
     csv_writer.writerow([_COL_SENTENCE, _COL_DCID])
-    for sentence in sorted(text2sv.keys()):
-      dcids_str = ';'.join(sorted(text2sv[sentence]))
-      texts.append(sentence)
-      dcids.append(dcids_str)
-      csv_writer.writerow([sentence, dcids_str])
+    for sentence in sentences:
+      csv_writer.writerow([sentence.text, sentence.dcid])
+
   # Write md5sum of preindex as a file
   with open(os.path.join(fm.local_output_dir(), _MD5_SUM_FILE), 'w') as f:
     f.write(get_md5sum(fm.preindex_csv_path()))
-  return texts, dcids
 
+  return sentences
 
-def compute_embeddings(texts: List[str],
-                       model: EmbeddingsModel) -> List[List[float]]:
-  """
-  Compute embeddings for a list of text strings.
-  """
+
+def retrieve_embeddings(
+    model: EmbeddingsModel,
+    target_sentences: List[SentenceObject],
+    saved_sentences: List[SentenceObject],
+) -> List[SentenceObject]:
+  """Compute embeddings for a list of sentence objects"""
```

Comment:
Describe the args and return value some more-- what is target and saved?  Do the input SentenceObjects have vector?

Reply:
done